### PR TITLE
[POC] create patch API object

### PR DESCRIPTION
**Pull request by deads2k**
Adds a patch API object that allows someone to create a patch inside of a template and have it applied like any other object you're creating.  It opens up new categories of templates that update existing objects to add new features and powers.

```yaml
apiVersion: v1
kind: Patch
spec:
  version: v1
  resource: serviceaccounts
  namespace: other
  name: default
  type: strategic
  patch: '{"metadata":{"labels":{"foo":"bar"}}}'
```
`oc create -f mypatch.yaml`

The new impersonation plus impersonation-scopes gave us the ability to do this.

@openshift/api-review review the API concept.  Clearly it needs more tests and some upstream fixes, because of course `apply` broke again. sigh

### CODE REVIEW

File: pkg/patch/registry/patch/proxy.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	mapping, err := r.restMapper.RESTMapping(targetKind.GroupKind(), targetKind.Version)
+	if err != nil {
+		return nil, kapierrors.NewBadRequest(fmt.Sprintf("cannot find mapping for resource %#v", targetResource))
+	}
+
+	patchRequest := restClient.Patch(kapi.StrategicMergePatchType).SetHeader(authenticationapi.ImpersonateUserHeader, user.GetName())
+	for _, scope := range user.GetExtra()[authorizationapi.ScopesKey] {
+		patchRequest = patchRequest.SetHeader(authenticationapi.ImpersonateUserScopeHeader, scope)
```
Comment (deads2k) [inline pkg/patch/registry/patch/proxy.go]: @liggitt I need "Impersonate-User-Group".....

**Comment by csrwng** [comment]
So just thinking out loud ... A patch would have to apply to something that already exists, which is useful when you have a set of objects that you know have been created before-hand (presumably by a different template). I'm not sure that I'd want the patch to apply to an object in the server ... rather I'd want it to apply before my template is created so we can de-dup some things. For example, right now we have ephemeral and persistent versions of templates which are a pain to maintain. Ideally we'd have a base template and an add-on patch to make it persistent. Normally I'd want that to be processed all at once on creation though.

@bparees thoughts?

**Comment by bparees** [comment]
@csrwng well i agree it'd be preferable to be able to do something like "oc process -f basetemplate.json persistenttemplate.json" and have all the patching performed before a single object is created, that's probably well beyond what this PR should cover.

So for purposes of this pr the flow would be:
oc process -f basetemplate.json
(stuff gets deployed)
oc process -f persistenttemplate.json
(stuff gets updated to have PVs and gets redeployed)

not ideal, but it's a step in the right direction, anyway.

**Comment by deads2k** [comment]
This is likely to be superceded by a bulk API.

**Comment by smarterclayton** [comment]
I think we may want this for creation anyway.  but we may want bulk patch.

**Pull request closed without merging**